### PR TITLE
[IMP] base, web: support custom x_active fields OoB

### PR DESCRIPTION
**Pull request by bouvyd**
Support active fields `x_active` like their equivalent `active` base fields:
- `active_test` context key should take them into account
- ditto  for archive/unarchive methods
- list/form views should support them OoB through the action menu

Small fix for importable modules while i'm there

This is the first brick in an incoming studio task that allows users to add per-defined behaviours on their new models. Archiving is one such behaviours and it requires this PR to work properly.

### CODE REVIEW

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 1]
```diff
@@ -3285,6 +3285,22 @@ var BasicModel = AbstractModel.extend({
         }
         return commands;
     },
+    /**
+     * For a given resource of type 'record', get the active field, if any.
+     *
+     * Since the ORM can support both `active` and `x_active` fields for
+     * the archiving mechanism, check if any such field exist and prioritize
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: exists

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 2]
```diff
@@ -3285,6 +3285,22 @@ var BasicModel = AbstractModel.extend({
...
+     * them. The `active` field should always take priority over its custom
+     * version.
+     *
+     * @param {Object} record local resource
+     * @returns {String|undefined} the field name to use for archiving purposes
+     *   ('active', 'x_active') or undefined if no such field is present
+     */
+    _getActiveField: function(record) {
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: should be public (and thus moved to the Public section)

Comment (aab-odoo) [reply addons/web/static/src/js/views/basic/basic_model.js]: function (record)

File: addons/web/static/src/js/views/form/form_controller.js [hunk 3]
```diff
@@ -519,8 +521,10 @@ var FormController = BasicController.extend({
             var unarchive_item = _.find(this.sidebar.items.other, function(item) {
                 return item.classname && item.classname.includes('o_sidebar_item_unarchive')
             })
+            const activeField = this.model._getActiveField(this.renderer.state);
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/form/form_controller.js]: Typically, we don't access the state of the renderer, so I'd suggest:
```js
const state = this.model.get(this.handle, {raw: true});
const activeField = this.model._getActiveField(state);
```
Or you could still use this.initialState as your are interested in an invariable property.

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 4]
```diff
@@ -3285,6 +3285,22 @@ var BasicModel = AbstractModel.extend({
...
+        const fields = Object.keys(record.fields);
+        return fields.find((f) => f === 'active' || f === 'x_active');
```
Comment (rco-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: What if fields contains both `active` and `x_active`?
The doc says it should return `active`, but the implementation does not seem deterministic.

Comment (rco-odoo) [reply addons/web/static/src/js/views/basic/basic_model.js]: By the way, using `find` seems inefficient to me. I suggest
```suggestion
        return ('active' in fields ? 'active' :
                'x_active' in fields ? 'x_active' :
                undefined);
```

File: addons/web/static/tests/views/basic_model_tests.js [hunk 5]
```diff
@@ -21,6 +22,7 @@ odoo.define('web.basic_model_tests', function (require) {
                         category: { string: "Category M2M", type: 'many2many', relation: 'partner_type' },
                         date: { string: "Date Field", type: 'date' },
                         reference: { string: "Reference Field", type: 'reference', selection: [["product", "Product"], ["partner_type", "Partner Type"], ["partner", "Partner"]] },
+                        x_active: { string: "Custom Active", type: 'boolean', default: true},
```
Comment (rco-odoo) [inline addons/web/static/tests/views/basic_model_tests.js]: Many JS implementations use the insertion order for keys in objects.  So, in order to check that `_getActiveField` returns the right one, you should put them in reverse order: `'x_active'`, `'active'`.

**Comment by rco-odoo** [comment]
@dbo-odoo I just implemented the `_active_name` feature on models.
It simplifies the way to deal with an active field, I think.
I also adapted more code that deals with the active field.

**Comment by dbo-odoo** [comment]
@rco-odoo 
As discussed last month, I have removed the generic implementation using `_active_name` since it would imply a rather large change in the web client. There is currently no mechanism to get information about a model, it would either have to be shoe-horned into the `load` call for the action or the read of the record, but that would be quite hacky for something so specific.

Regarding the default value, well in my branch (which sets this field automatically python-side at model creation), I create an `ir.default` record to set it to True by default. Works well enough, even if it means adding support for archiving manually is a bit more complex.

**Comment by dbo-odoo** [comment]
After some backchanneling with @rco-odoo, the final solution will be to use an `_active_name` attribute on the model which specifies the field name to use for archiving purposes. Since the client does not have an elegant way to handle any name on this field, the ORM restricts this attribute to either `active` or `x_active`.

This should be merged soon as we have a consensus now :)

@aab-odoo you can have a last look if you wish to see the changes following your last review.

**Comment by dbo-odoo** [comment]
@aab-odoo I'll merge tomorrow, you have <24h to speak now or forever hold your peace :)

**Comment by dbo-odoo** [comment]
Jean-Luc Picard saying 'make it so' whilst waving his finger
@robodoo r+ rebase-ff (my last push was just a rebase and commit msg change, so it should still be green)

**Pull request closed without merging**